extensions-contrib/moving-average-query/src/main/java/org/apache/druid/query/movingaverage/MovingAverageIterable.java

```diff
@@ -0,0 +1,302 @@
+        List<DimensionSpec> dims,
+        List<AveragerFactory<?, ?>> factories,
+        Map<String, Object> fakeEvents,
+        Map<String, AggregatorFactory> aggMap
+    )
+    {
+      this.dims = dims;
+      this.factories = factories;
+      this.fakeEvents = fakeEvents;
+      this.aggMap = aggMap;
+
+      yielder = rows.toYielder(null, new IdentityYieldingAccumulator<RowBucket>());
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#hasNext()
+     */
+    @Override
+    public boolean hasNext()
+    {
+
+      if (saveNext != null) {
+        return true;
+      }
+
+      try {
+        saveNext = internalNext();
+        return true;
+      }
+      catch (NoSuchElementException e) {
+        return false;
+      }
+
+
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#next()
+     */
+    @Override
+    public Row next()
+    {
+
+      if (saveNext != null) {
+        Row retVal = saveNext;
+        saveNext = null;
+        return retVal;
+      }
+
+      return internalNext();
+    }
+
+    private Row internalNext()
+    {
+      if (cache == null && !yielder.isDone()) {
+        cache = yielder.get();
+        yielder = yielder.next(cache);
+
+        cacheIter = cache.getRows().iterator();
+      }
+
+      Row r;
+
+      // return rows from the cached RowBucket
+      if (cacheIter != null) {
+        if (cacheIter.hasNext()) {
+          r = cacheIter.next();
+          Map<String, Object> key = MovingAverageHelper.getDimKeyFromRow(dims, r);
+          seenKeys.add(key);
+          r = computeMovingAverage(key, r, false);
+          if (r != null) {
+            return r;
+          } else {
+            throw new NoSuchElementException();
+          }
+        } else {
+          Set<Map<String, Object>> averagerKeys = new HashSet<>(averagers.keySet());
+          averagerKeys.removeAll(seenKeys);
+          averagersKeysIter = averagerKeys.iterator();
+          cacheIter = null;
+        }
+      }
+
+      // return fake rows for unseen dimension combinations
+      if (averagersKeysIter != null) {
+        while (averagersKeysIter.hasNext()) {
+          Map<String, Object> dims = averagersKeysIter.next();
+          Map<String, Object> fakeEventsCopy = new HashMap<>(fakeEvents);
+
+          dims.forEach((dim, value) -> {
+            fakeEventsCopy.put(dim, value);
+          });
+
+          r = computeMovingAverage(dims, new MapBasedRow(cache.getDateTime(), fakeEventsCopy), true);
+          if (r != null) {
+            return r;
+          }
+        }
+
+        seenKeys.clear();
+        averagersKeysIter = null;
+        cache = null;
+      }
+
+      if (cacheIter == null && yielder.isDone()) {
+        // we should never get here. For some reason, there is
+        // no more work to do, so continuing to iterate will infinite loop
+        throw new NoSuchElementException();
+      }
+
+      // nothing to do here, so move on to the next row
+      return internalNext();
```

Comment:
This recursion can incur stack overflow if there are a lot of empty buckets.

Reply:
I changed to iteration (`while (true)`). The safety point is still returning null when RowBucketIterable is exhausted.